### PR TITLE
[baremetal] tinycrypt ARM assembler and other optimisations

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR supersedes https://github.com/ARMmbed/mbedtls/pull/3750 by introducing an additional optional optimization flag. Original description by @kjbracey-arm (author of most of the code in this PR):

> - Add T32/A32 assembler optimisations for core tinycrypt/microecc routines, for ARMC5 and GCC.
> - Use fast integer types to avoid frequent narrowing instructions
> - Use __builtin_clz
> - Avoid boolean ops (speed optimisation, and to achieve constant-time intent. Boolean operations are generally not compiled as constant time on T32/A32).
> 
> Achieves something close to a doubling of speed of DTLS handshake on an M0+/M4 pair of devices.
>

### CODE REVIEW

File: configs/baremetal.h [hunk 1]
```diff
@@ -116,6 +116,8 @@
     MBEDTLS_SSL_EXTENDED_MS_ENFORCE_ENABLED
 
 #define MBEDTLS_USE_TINYCRYPT
+#define MBEDTLS_HAVE_ASM
```
Comment (shelib01) [inline configs/baremetal.h:119]: @AndrzejKurek , why do we need MBEDTLS_HAVE_ASM flag in baremetal?

Comment (shelib01) [reply configs/baremetal.h:119]: I think it is already defined in main config and we should not touch it ...

Comment (AndrzejKurek) [reply configs/baremetal.h:119]: What do you mean by the "main config"? I assumed that we want to have the ASM optimization on at all times, so ASM has to be enabled. I can turn the logic around if you want, to check if ASM is enabled, and if so, then turn on the ASM optimization instead.

Comment (shelib01) [reply configs/baremetal.h:119]: I mean that in this file https://github.com/ARMmbed/mbedtls/blob/baremetal/include/mbedtls/config.h#L61 it is enabled by default. 
Actually I am asking why do we want to perform coupling between  2 flags? Why only we can not switch on only MBEDTLS_OPTIMIZE_TINYCRYPT_ASM ?

Comment (AndrzejKurek) [reply configs/baremetal.h:119]: Because it won't work without tinycrypt and MBEDTLS_HAVE_ASM turned on.

File: configs/baremetal.h [hunk 2]
```diff
@@ -117,8 +117,9 @@
 
 #define MBEDTLS_USE_TINYCRYPT
 #define MBEDTLS_HAVE_ASM
-#define MBEDTLS_OPTIMIZE_TINYCRYPT_ASM
-
+#if !( defined(__STRICT_ANSI__) && defined(__CC_ARM) )
+    #define MBEDTLS_OPTIMIZE_TINYCRYPT_ASM
```
Comment (shelib01) [inline configs/baremetal.h:121]: @AndrzejKurek not sure I understand : the optimization will always switched on if cc_arm defined ? Why we cannot allow to user decide if use it or not ?

Comment (shelib01) [reply configs/baremetal.h:121]: @AndrzejKurek , got the point, it is fine

File: tinycrypt/ecc.c [hunk 3]
```diff
@@ -203,14 +210,177 @@ int uECC_curve_public_key_size(void)
...
+#else
+    MOVS    r1,#0
+    STRD    r1,r1,[r0,#0]       // Only Thumb2 STRD can store same reg twice, not ARM
+    STRD    r1,r1,[r0,#8]
+    STRD    r1,r1,[r0,#16]
+    STRD    r1,r1,[r0,#24]
+    BX      lr
+#endif
+}
+#elif defined MBEDTLS_OPTIMIZE_TINYCRYPT_ASM && defined __GNUC__ && defined __arm__
```
Comment (shelib01) [inline tinycrypt/ecc.c:236]: @AndrzejKurek , @kjbracey-arm , as the flags defined right now , it seems this #elif is never stepped in . In baremetal.h defined 

```
#if !( defined(__STRICT_ANSI__) && defined(__CC_ARM) )
    #define MBEDTLS_OPTIMIZE_TINYCRYPT_ASM
```
and in this file 
```

#if defined MBEDTLS_OPTIMIZE_TINYCRYPT_ASM && defined __CC_ARM
.
.
.
#elif defined MBEDTLS_OPTIMIZE_TINYCRYPT_ASM && defined __GNUC__ && defined __arm__
```
What are we trying to achieve here ?

File: configs/baremetal.h (hunk 2 above)
Comment (AndrzejKurek) [reply configs/baremetal.h:121]: |         | STRICT   | !STRICT |
|---------|----------|---------|
| CC_ARM  | NO OPTIM | OPTIM   |
| !CC_ARM | OPTIM    | OPTIM   |

File: tinycrypt/ecc.c (hunk 3 above)
Comment (shelib01) [reply tinycrypt/ecc.c:236]: Understood it

**Review by shelib01 — APPROVED** [review]

**Pull request merged**